knative-operator/pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,186 @@
+}
+
+// TryStartTelemetry setups telemetry per component either Eventing, KnativeKafka or Serving.
+// When called it assumes that the component has status ready.
+func (t *Telemetry) TryStartTelemetry(api client.Client, mgr manager.Manager) error {
+	log.Info("starting telemetry for:", "component", t.name)
+	// Initialize metrics before we start the corresponding controller.
+	// There is a tiny window to miss events here, but should be ok for telemetry purposes.
+	t.InitializeAndTakeMetricsSnapshot(api)
+	// Start our controller in a goroutine so that we do not block.
+	if t.tc != nil { // meant to allow nop objects in tests
+		go func() {
+			// Block until our controller manager is elected leader. We presume our
+			// entire process will terminate if we lose leadership, so we don'telemetry need
+			// to handle that.
+			<-mgr.Elected()
+			// Start our controller. This will block until it is stopped
+			// or the controller returns an error.
+			if err := (*t.tc).Start(t.stop); err != nil {
+				log.Error(err, "cannot start telemetry controller for", "component", t.name)
+			}
+		}()
+	}
+	return nil
+}
+
+// TryStopTelemetry stops telemetry per component either Eventing, KnativeKafka or Serving
+// When called it assumes that we are reconciling a deletion event.
+func (t *Telemetry) TryStopTelemetry() {
+	log.Info("stopping telemetry for:", "component", t.name)
+	// Stop the telemetry controller
+	// the lock above makes sure we close once and not panic
+	close(t.stop)
+	// Can'telemetry use a closed channel
+	t.stop = make(chan struct{})
+	// Remove snapshot entries for the components so that snapshot is does not get
+	// unbounded since the telemetry controller can be restarted multiple times.
+	t.metricsSnapshot.Range(func(key interface{}, value interface{}) bool {
+		t.metricsSnapshot.Delete(key)
+		return true
+	})
+}
+
+// InitializeAndTakeMetricsSnapshot is used for taking a global snapshot of metrics
+// before we start any telemetry controller. If the operator is restarted
+// client metrics will not reflect the state in the cluster so there is need to restore current state.
+// It should run each time operator is initialized and for each trial to start a Telemetry controller.
+// Snapshot is used for skipping events already counted but also received later after
+// telemetry controllers are started. Metrics are initialized here and kept around as long as the
+// Serverless Operator is running. No metric label is removed when components are removed so that
+// if any component eg. Eventing is removed and CRDs exist, it will show existing CRs correctly.
+// Telemetry should be used in the background so no error is returned here.
+func (t *Telemetry) InitializeAndTakeMetricsSnapshot(api client.Client) {
+	switch t.name {
+	case "eventing":
+		sourcesG = serverlessTelemetryG.WithLabelValues("source")
+		pingSourceList := &eventingsourcesv1beta1.PingSourceList{}
+		if err := api.List(context.TODO(), pingSourceList); err == nil {
```

Comment:
I wonder if we should just always use a List operation to gather metrics then. AFAIK, they should be coming from cache anyway (with the `Watch` being setup), so what if we don't do an `Add/Dec` operation on individual changes but instead do a `Set` operation on each individual change by listing everything. If everything comes from cache, there's 0 cost to that and it'd make the "snapshot" obsolete.

Reply:
The problem is that for stuff that are not in cache yet you will do that call multiple times for each type of resource you get an event for. The reconciler uses c-r client to list stuff, and it touches the cache as you mention. 
The other problem is that if you have no event coming in to trigger the controller because someone restarted the operator pod you will get a zero value for all metrics. T

Reply:
But shouldn't we get a global resync on controller startup?

Reply:
The complete solution would be to do a global resync there as well. In the code I have this comment:

```
	// Initialize metrics before we start the corresponding controller.
	// There is a tiny window to miss events here, but should be ok for telemetry purposes.
``
At this point we take the snapshot and then start the controller. I dont expect to lose a considerable number of events in between.
For telemetry I am going to use something like `avg by (type) (avg_over_time(serverless_telemetry[30m]))`, so I dont expect much of a difference in general. I can also do that for the sake of correctness. WDYTH?

Reply:
My point is: Why doesn't it already do a global resync anyway? AFAIK starting a watcher does that by design.

Reply:
Ok will do.